File: services/paper_ingestion/paper_ingestion/models/notes.py
"""Paper-note Pydantic models."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class NoteCreate(BaseModel):
    """Request body for creating a paper note."""

    user_note: str = Field(..., min_length=1, max_length=5000)
    highlight_text: str | None = None
    page_number: int | None = Field(default=None, ge=1)


class NoteUpdate(BaseModel):
    """Request body for updating a paper note."""

    user_note: str | None = Field(default=None, min_length=1, max_length=5000)
    highlight_text: str | None = None
    page_number: int | None = Field(default=None, ge=1)


class NoteResponse(BaseModel):
    """Response for a paper note."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    paper_id: int
    user_note: str
    highlight_text: str | None = None
    page_number: int | None = None
    source: Literal["user", "zotero"] = "user"
    zotero_annotation_key: str | None = None
    verification_status: Literal["unverified", "verified", "failed"] = "unverified"
    verified_quote: str | None = None
    verified_page_number: int | None = None
    promoted_at: datetime | None = None
    stale: bool = False
    created_at: datetime

File: frontend/src/types/notes.ts
export interface Note {
  id: number;
  paper_id: number;
  user_note: string;
  highlight_text: string | null;
  page_number: number | null;
  source: 'user' | 'zotero';
  zotero_annotation_key: string | null;
  verification_status: 'unverified' | 'verified' | 'failed';
  verified_quote: string | null;
  verified_page_number: number | null;
  promoted_at: string | null;
  created_at: string;
}

// --- Spatial PDF highlights (in-PDF annotation reader) ---

/** Normalized, top-origin rectangle on a PDF page (coordinates in [0, 1]). */
export interface Rect {
  x0: number;
  y0: number;
  x1: number;
  y1: number;
}

/** react-pdf-highlighter geometry: union box plus its per-line rectangles. */
export interface HighlightRect {
  boundingRect: Rect;
  rects: Rect[];
}

export interface Highlight {
  id: number;
  paper_id: number;
  page: number;
  rect: HighlightRect;
  note: string | null;
  color: string | null;
  quote: string | null;
  created_at: string;
}

export interface HighlightCreate {
  page: number;
  rect: HighlightRect;
  note?: string | null;
  color?: string | null;
  quote?: string | null;
}

export interface HighlightUpdate {
  note?: string | null;
  color?: string | null;
}
